### PR TITLE
add reader for UPAK lites

### PR DESCRIPTION
**Pull request by xgess**
Adds a basic reader for the new UpkLite. 

this PR also implements `CORE-9007`, adding resets. I put it here because we were still blocked on the writer, and it seemed silly to have the nested PRs when it's all pretty interrelated.

here's a link to the branch that uses this reader as part of the team load: https://github.com/keybase/client/pull/14028

noteworthy caveats:
* The tests for this new UPKLite do not verify correctness, only that everything matches with the load of a full UPAK. 
* We have not yet implemented caching explicitly at the chain level. There's some that happens already at the link level. We're planning on doing that in the next sprint, and it really *shouldn't* affect getting this into the team loader, at least as a final fallback option. even if, in most circumstances right now, this would not be a preferable way to load a user.

### CODE REVIEW

File: go/libkb/sig_chain.go [hunk 1]
```diff
@@ -669,16 +674,16 @@ func (sc *SigChain) verifySubchain(m MetaContext, kf KeyFamily, links ChainLinks
 				// So that the server can't roll back someone's active wallet address.
 				return cached, cki, SigchainV2StubbedDisallowed{}
 			}
-			sc.G().VDL.Log(VLog1, "| Skipping over stubbed-out link: %s", link.id)
+			m.G().VDL.Log(VLog1, "| Skipping over stubbed-out link: %s", link.id)
```
Comment (maxtaco) [inline go/libkb/sig_chain.go]: `m.VLogf`

File: go/libkb/sig_chain.go [hunk 2]
```diff
@@ -669,16 +674,16 @@ func (sc *SigChain) verifySubchain(m MetaContext, kf KeyFamily, links ChainLinks
...
 		}
 
 		tcl, w := NewTypedChainLink(link)
 		if w != nil {
-			w.Warn(sc.G())
+			w.Warn(m.G())
 		}
 
-		sc.G().VDL.Log(VLog1, "| Verify link: %s", link.id)
+		m.G().VDL.Log(VLog1, "| Verify link: %s", link.id)
```
Comment (maxtaco) [inline go/libkb/sig_chain.go]: `m.VLogf`

File: go/libkb/sig_chain.go [hunk 3]
```diff
@@ -696,7 +701,7 @@ func (sc *SigChain) verifySubchain(m MetaContext, kf KeyFamily, links ChainLinks
 		isModifyingKeys := isDelegating || tcl.Type() == string(DelegationTypePGPUpdate)
 		isFinalLink := (linkIndex == len(links)-1)
 		hasRevocations := link.HasRevocations()
-		sc.G().VDL.Log(VLog1, "| isDelegating: %v, isModifyingKeys: %v, isFinalLink: %v, hasRevocations: %v",
+		m.G().VDL.Log(VLog1, "| isDelegating: %v, isModifyingKeys: %v, isFinalLink: %v, hasRevocations: %v",
```
Comment (maxtaco) [inline go/libkb/sig_chain.go]: `m.VLogf`

File: go/libkb/upak_lite.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+	}
+	// compute keys
+	err = l.VerifySigsAndComputeKeys()
+
+	return l.chain, nil
+}
+
+func (l *HighSigChainLoader) selfUID() (uid keybase1.UID) {
+	//for now let's always assume this isn't applicable, but we can add it later
+	return
```
Comment (maxtaco) [inline go/libkb/upak_lite.go]: `return uid`

File: go/libkb/upak_lite.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+}
+
+func (l *HighSigChainLoader) LoadFromServer() (err error) {
+	srv := l.GetMerkleTriple()
+	l.dirtyTail, err = l.chain.LoadFromServer(l.M(), srv, l.selfUID())
+	return
```
Comment (maxtaco) [inline go/libkb/upak_lite.go]: `return err`

File: protocol/avdl/keybase1/upk.avdl [hunk 6]
```diff
@@ -111,4 +111,26 @@ protocol UPK {
...
+  record UpkLitev1 {
+      UID uid;
+      string username;
+      Seqno eldestSeqno;
+      StatusCode status;
+      // Sorted by generation ascending
+      map<KID, PublicKeyV2NaCl> deviceKeys;
+  }
+
+  record UpkLitev1AllIncarnations {
```
Comment (maxtaco) [inline protocol/avdl/keybase1/upk.avdl]: Isn't `V` capital on the heavy version? Follow whatever template we have there.

File: go/libkb/upak_lite.go [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+	loader := NewHighSigChainLoader(m, user, leaf)
+	highChain, err := loader.Load()
+	if err != nil {
+		return nil, err
+	}
+	return highChain.ToUPAKLite(user)
+}
+
+type HighSigChainLoader struct {
+	BaseSigChainLoader
```
Comment (maxtaco) [inline go/libkb/upak_lite.go]: Does this buy you anything?  I'm not sure what it means to contain an anonymous interface.  My guess is this is a noop.

File: go/client/cmd_upak.go [hunk 8]
```diff
@@ -44,11 +49,23 @@ func (c *cmdUPAK) Run() error {
...
 	if err = RegisterProtocolsWithContext(nil, c.G()); err != nil {
 		return err
 	}
 
+	if c.lite {
+		upakLite, err := userClient.GetUPAKLite(context.Background(), c.uid)
+		if err != nil {
+			return err
+		}
+		jsonOut, err := json.Marshal(upakLite)
```
Comment (heronhaye) [inline go/client/cmd_upak.go]: MarshalIndent might be nicer for debugging

File: go/libkb/chain_link.go [hunk 9]
```diff
@@ -205,18 +205,19 @@ var badWhitespaceChainLinks = map[keybase1.SigID]string{
...
-	dirty            bool
-	revocationsCache *[]keybase1.SigID
-	computedHighSkip *HighSkip
+	parent            *SigChain
+	id                LinkID
+	hashVerified      bool
+	sigVerified       bool
+	payloadVerified   bool
+	chainVerified     bool
+	highChainVerified bool
```
Comment (heronhaye) [inline go/libkb/chain_link.go]: does this mean everything <= this seqno on the highchain is verified?

File: go/libkb/upak_lite.go [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+	"github.com/buger/jsonparser"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+)
+
+func LoadUPAKLite(arg LoadUserArg) (ret *keybase1.UpkLitev1AllIncarnations, err error) {
+	uid := arg.uid
+	m := arg.m
+
+	// get sig hints in order to populate during merkle leaf lookup
+	sigHints, err := LoadSigHints(m, arg.uid)
```
Comment (heronhaye) [inline go/libkb/upak_lite.go]: could probably get rid of this and put nil in `lookupMerkleLeaf`

File: go/libkb/upak_lite.go [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+	dirtyTail *MerkleTriple
+}
+
+type HighSigChain struct {
+	BaseSigChain
+	Contextified
+	uid                  keybase1.UID
+	username             NormalizedUsername
+	chainLinks           ChainLinks
+	localCki             *ComputedKeyInfos
```
Comment (heronhaye) [inline go/libkb/upak_lite.go]: will we need these `local` attrs if we're only going to use highsigchain for other users?

File: go/libkb/upak_lite.go [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+				// fallback to normal prevs if the link doesn't have a high_skip
+				expectedPrevSeqno = curr.GetSeqno() - 1
+				expectedPrevID = curr.GetPrev()
+			}
+			if !prev.id.Eq(expectedPrevID) {
+				return ChainLinkPrevHashMismatchError{fmt.Sprintf("Chain mismatch at seqno=%d", curr.GetSeqno())}
+			}
+			if prev.GetSeqno() != expectedPrevSeqno {
+				return ChainLinkWrongSeqnoError{fmt.Sprintf("Chain seqno mismatch at seqno=%d (previous=%d)", curr.GetSeqno(), prev.GetSeqno())}
+			}
```
Comment (heronhaye) [inline go/libkb/upak_lite.go]: could also check for 0 seqno/null id for the first link

File: go/libkb/upak_lite.go (hunk 5 above)
Comment (xgess) [reply go/libkb/upak_lite.go]: 👍

File: protocol/avdl/keybase1/upk.avdl (hunk 6 above)
Comment (xgess) [reply protocol/avdl/keybase1/upk.avdl]: oops. fixed.

File: go/libkb/sig_chain.go (hunk 1 above)
Comment (xgess) [reply go/libkb/sig_chain.go]: 👍

File: go/libkb/sig_chain.go (hunk 2 above)
Comment (xgess) [reply go/libkb/sig_chain.go]: 👍

File: go/libkb/sig_chain.go (hunk 3 above)
Comment (xgess) [reply go/libkb/sig_chain.go]: 👍

File: go/libkb/upak_lite.go (hunk 4 above)
Comment (xgess) [reply go/libkb/upak_lite.go]: 👍

File: go/client/cmd_upak.go (hunk 8 above)
Comment (xgess) [reply go/client/cmd_upak.go]: cool. i'm not super opinionated; i prefer to pipe it to jq either way.

File: go/libkb/upak_lite.go (hunk 11 above)
Comment (xgess) [reply go/libkb/upak_lite.go]: i believe you are correct. will remove this.

File: go/client/cmd_upak.go [hunk 13]
```diff
@@ -44,11 +49,23 @@ func (c *cmdUPAK) Run() error {
...
+	if c.lite {
+		upakLite, err := userClient.GetUPAKLite(context.Background(), c.uid)
+		if err != nil {
+			return err
+		}
+		jsonOut, err := json.MarshalIndent(upakLite, "", "  ")
+		if err != nil {
+			return err
+		}
+		c.G().UI.GetTerminalUI().Output(string(jsonOut) + "\n")
```
Comment (heronhaye) [inline go/client/cmd_upak.go]: curious, do you know if this is escaped for malicious unicode strings? I don't think there's any arbitrary user input for upak lites so it's okay here regardless.

File: go/libkb/upak_lite.go [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+	}
+	var links ChainLinks
+	var lastLink *ChainLink
+
+	jsonparser.ArrayEach(body, func(value []byte, dataType jsonparser.ValueType, offset int, inErr error) {
+		var link *ChainLink
+
+		parentSigChain := &SigChain{} // because we don't want the cache to use these
+		link, err = ImportLinkFromServer(m.G(), parentSigChain, value, selfUID)
+		links = append(links, link)
```
Comment (heronhaye) [inline go/libkb/upak_lite.go]: preallocating the array could be a perf boost, but could wait until profiling to do it

File: go/client/cmd_upak.go (hunk 13 above)
Comment (xgess) [reply go/client/cmd_upak.go]: dunno

File: go/libkb/upak_lite.go (hunk 14 above)
Comment (xgess) [reply go/libkb/upak_lite.go]: +1

File: go/libkb/upak_lite.go (hunk 12 above)
Comment (xgess) [reply go/libkb/upak_lite.go]: this is done.

**Comment by xgess** [comment]
@maxtaco i think this is ready for a re-review at your leisure

File: go/libkb/upak_lite.go [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+	}
+	return &deviceKeys
+}
+
+func buildUpkLiteAllIncarnations(m MetaContext, user *User, leaf *MerkleUserLeaf, hsc *HighSigChain) (ret *keybase1.UpkLiteV1AllIncarnations, err error) {
+	// build the current UPKLiteV1
+	uid := user.GetUID()
+	name := user.GetName()
+	status := user.GetStatus()
+	eldestSeqno := hsc.chainLinks[0].GetSeqno()
```
Comment (xgess) [inline go/libkb/upak_lite.go]: surya says: this isn't right

File: go/libkb/upak_lite.go [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+}
+
+func (hsc *HighSigChain) LoadFromServer(m MetaContext, t *MerkleTriple, selfUID keybase1.UID) (dirtyTail *MerkleTriple, err error) {
+	// get the high sigs from the server
+	// ------------------
+	m, tbs := m.WithTimeBuckets()
+
+	apiArg := APIArg{
+		Endpoint:    "sig/get_high",
+		SessionType: APISessionTypeREQUIRED,
```
Comment (heronhaye) [inline go/libkb/upak_lite.go]: this should be `APISessionTypeOptional` so you can load skipchains when logged out

Comment (heronhaye) [reply go/libkb/upak_lite.go]: (don't worry about the logged-out test)

File: go/libkb/upak_lite.go (hunk 15 above)
Comment (maxtaco) [reply go/libkb/upak_lite.go]: Why not?

File: go/libkb/upak_lite.go (hunk 16 above)
Comment (xgess) [reply go/libkb/upak_lite.go]: done. and with a test. :)

File: go/libkb/upak_lite.go (hunk 15 above)
Comment (xgess) [reply go/libkb/upak_lite.go]: fixed

File: go/libkb/upak_lite.go [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"github.com/buger/jsonparser"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+)
+
+func LoadUPAKLite(arg LoadUserArg) (ret *keybase1.UpkLiteV1AllIncarnations, err error) {
```
Comment (heronhaye) [inline go/libkb/upak_lite.go]: maybe this should be private function if you expect to only call it through a regular CachedUPAKLoader

File: go/libkb/upak_lite.go (hunk 15 above)
Comment (xgess) [reply go/libkb/upak_lite.go]: @maxtaco: resets

File: go/libkb/upak_lite.go [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+		return nil, err
+	}
+	dirtyTail = lastLink.ToMerkleTriple()
+
+	hsc.chainLinks = links
+	return dirtyTail, err
+}
+
+func (hsc *HighSigChain) VerifyChain(m MetaContext) (err error) {
+	m.CDebugf("+ HighSigChain#VerifyChain()")
```
[the remark this answers is not in the document]
Comment (xgess) [reply go/libkb/upak_lite.go]: 👍

File: go/libkb/upak_lite.go [hunk 19]
```diff
@@ -0,0 +1,359 @@
...
+			var expectedPrevSeqno keybase1.Seqno
+			if curr.GetHighSkip() != nil {
+				expectedPrevSeqno = curr.GetHighSkip().Seqno
+				expectedPrevID = curr.GetHighSkip().Hash
+			} else {
+				// fallback to normal prevs if the link doesn't have a high_skip
+				expectedPrevSeqno = curr.GetSeqno() - 1
+				expectedPrevID = curr.GetPrev()
+			}
+			if i == 0 && (expectedPrevSeqno != 0 || expectedPrevID != nil) {
```
[the remark this answers is not in the document]
Comment (xgess) [reply go/libkb/upak_lite.go]: brain fart. i added this check after the main thread. will fix.

File: protocol/avdl/keybase1/upk.avdl [hunk 20]
```diff
@@ -111,4 +111,27 @@ protocol UPK {
       case V1: UserPlusAllKeys;
       case V2: UserPlusKeysV2AllIncarnations;
   }
+
+  enum UpkLiteMinorVersion {
```
[the remark this answers is not in the document]
Comment (xgess) [reply protocol/avdl/keybase1/upk.avdl]: 👍

**Comment by xgess** [comment]
@maxtaco another ping here. thanks.

**Pull request merged**